server/server.go

```diff
@@ -551,8 +557,49 @@ func (a *ArgoCDServer) newHTTPServer(ctx context.Context, port int, grpcWebHandl
 	// we use our own Marshaler
 	gwMuxOpts := runtime.WithMarshalerOption(runtime.MIMEWildcard, new(jsonutil.JSONMarshaler))
 	gwCookieOpts := runtime.WithForwardResponseOption(a.translateGrpcCookieHeader)
-	gwmux := runtime.NewServeMux(gwMuxOpts, gwCookieOpts)
-	mux.Handle("/api/", gwmux)
+
+	// This function adds the X-CSRF-Token header to the HTTP response so the browser can supply it
+	// with POST/PUT/DELETE requests.
+	//
+	// gorilla-csrf uses context from a http.Request object, so we have to supply an empty request
+	// with the current context set in our grpc-gateway setup for gorilla to create a proper token.
+	csrfAddResponseHeaderFunc := func(ctx context.Context, w http.ResponseWriter, resp golang_proto.Message) error {
+		r := http.Request{}
+		token := csrf.Token(r.WithContext(ctx))
+		if token != "" {
+			w.Header().Set("X-CSRF-Token", token)
+		}
+		return nil
+	}
+
+	gwForwarder := runtime.WithForwardResponseOption(csrfAddResponseHeaderFunc)
+
+	// Our HTTP JSON API is CSRF protected, so we add Gorilla CSRF middleware
+	//
+	// Client must either send the "Authorization" header (API usage from non-browser clients) or a valid
+	// CSRF token in "X-CSRF-Token" header to pass CSRF protection middleware.
+	//
+	var gwmux *runtime.ServeMux
+
+	if a.ArgoCDServerOpts.DisableCsrf {
+		log.Warnf("CSRF protection is disabled -- do not use this in production")
+		gwmux = runtime.NewServeMux(gwMuxOpts, gwCookieOpts)
+		mux.Handle("/api/", gwmux)
+	} else {
+		// We need 32-byte non-changing key, so we just re-use the key used to sign JWT
+		gwmux = runtime.NewServeMux(gwMuxOpts, gwCookieOpts, gwForwarder)
+		if a.settings.CsrfKey != nil && len(a.settings.CsrfKey) == 32 {
+			csrfKey := a.settings.CsrfKey
+			protmux := csrf.Protect(csrfKey,
+				csrf.Secure(a.useTLS()),
+				csrf.Path("/api/v1"),
+			)(gwmux)
+			mux.Handle("/api/", protmux)
+		} else {
+			panic("server.secretkey must be at least 32 byte long -- consider another value or --disable-csrf")
```

Comment:
Is this error message correct? It mentions server.secretkey, I would expect server.csrfkey.

Reply:
Another good catch, thanks!

Reply:
Also, error/panic message should state "exactly 32 byte long" now.